constraint/pkg/client/e2e_test.go

```diff
@@ -0,0 +1,694 @@
+			if err != nil {
+				t.Fatalf("got Review: %v", err)
+			}
+
+			got := rsps.Results()
+			want := []*types.Result{{
+				Constraint:        cstr,
+				Msg:               denied,
+				EnforcementAction: "deny",
+			}}
+
+			if diff := cmp.Diff(want, got, cmpopts.IgnoreFields(types.Result{},
+				"Metadata", "Review", "Resource")); diff != "" {
+				t.Error(diff)
+			}
+		})
+	}
+}
+
+func TestE2EAudit(t *testing.T) {
+	for _, tc := range denyAllCases {
+		t.Run(tc.name, func(t *testing.T) {
+			c, err := newTestClient()
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			ctx := context.Background()
+			_, err = c.AddTemplate(ctx, newConstraintTemplate("Foo", tc.rego, tc.libs...))
+			if err != nil {
+				t.Fatalf("got AddTemplate: %v", err)
+			}
+			cstr := newConstraint("Foo", "ph", nil, nil)
+			if _, err := c.AddConstraint(ctx, cstr); err != nil {
+				t.Fatalf("got AddConstraint: %v", err)
+			}
+			obj := &targetData{Name: "Sara", ForConstraint: "Foo"}
+			if _, err := c.AddData(ctx, obj); err != nil {
+				t.Fatalf("got AddData: %v", err)
+			}
+			rsps, err := c.Audit(ctx)
+			if err != nil {
+				t.Fatalf("got Audit error: %v, want nil", err)
+			}
+
+			got := rsps.Results()
+			want := []*types.Result{{
+				Constraint:        cstr,
+				Msg:               denied,
+				EnforcementAction: "deny",
+				Resource:          obj,
+			}}
+
+			if diff := cmp.Diff(want, got, cmpopts.IgnoreFields(types.Result{},
+				"Metadata", "Review")); diff != "" {
+				t.Error(diff)
+			}
+		})
+	}
+}
+
+func TestE2EAuditX2(t *testing.T) {
+	for _, tc := range denyAllCases {
+		t.Run(tc.name, func(t *testing.T) {
+			c, err := newTestClient()
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			ctx := context.Background()
+			_, err = c.AddTemplate(ctx, newConstraintTemplate("Foo", tc.rego, tc.libs...))
+			if err != nil {
+				t.Fatalf("got AddTemplate: %v", err)
+			}
+			cstr := newConstraint("Foo", "ph", nil, nil)
+			if _, err := c.AddConstraint(ctx, cstr); err != nil {
+				t.Fatalf("got AddConstraint: %v", err)
+			}
+			obj := &targetData{Name: "Sara", ForConstraint: "Foo"}
+			if _, err := c.AddData(ctx, obj); err != nil {
+				t.Fatalf("got AddData: %v", err)
+			}
+			obj2 := &targetData{Name: "Max", ForConstraint: "Foo"}
+			if _, err := c.AddData(ctx, obj2); err != nil {
+				t.Fatalf("got AddDataX2: %v", err)
+			}
+			rsps, err := c.Audit(ctx)
+			if err != nil {
+				t.Fatalf("got Audit: %v", err)
+			}
+
+			got := rsps.Results()
+			want := []*types.Result{{
+				Constraint:        cstr,
+				EnforcementAction: "deny",
+				Msg:               denied,
+			}, {
+				Constraint:        cstr,
+				EnforcementAction: "deny",
+				Msg:               denied,
+			}}
+
+			if diff := cmp.Diff(want, got, cmpopts.IgnoreFields(types.Result{},
+				"Metadata", "Review", "Resource")); diff != "" {
+				t.Error(diff)
+			}
+		})
+	}
+}
+
+func TestE2EAutoreject(t *testing.T) {
```

Comment:
How does `denied` compare to `rejection`?  Some comments above this test (and perhaps others) would be helpful for someone understanding the different usage patterns exercised by these various tests.

Reply:
Good point. I'll add comments